x/stakeibc/keeper/hooks.go

```diff
@@ -239,107 +237,106 @@ func (k Keeper) SetWithdrawalAddress(ctx sdk.Context) {
-			amt, err := sdk.ParseCoinNormalized(processAmount)
-			if err != nil {
-				k.Logger(ctx).Error(fmt.Sprintf("Could not process coin %v: %v", hostZone.HostDenom, err.Error()))
-				return
-			}
-			err = k.DelegateOnHost(ctx, hostZone, amt, depositRecord.Id)
-			if err != nil {
-				k.Logger(ctx).Error(fmt.Sprintf("Did not stake %s on %s | err: %s", processAmount, hostZone.ChainId, err.Error()))
-				return
-			} else {
-				k.Logger(ctx).Info(fmt.Sprintf("Successfully submitted stake for %s on %s", processAmount, hostZone.ChainId))
-			}
+	maxDepositRecordsToStake := utils.Min(len(stakeDepositRecords), cast.ToInt(k.GetParam(ctx, types.KeyMaxStakeICACallsPerEpoch)))
+	k.Logger(ctx).Info(fmt.Sprintf("Staking %d out of %d deposit records", maxDepositRecordsToStake, len(stakeDepositRecords)))
+
+	for _, depositRecord := range stakeDepositRecords[:maxDepositRecordsToStake] {
+		k.Logger(ctx).Info(fmt.Sprintf("\t[StakeExistingDepositsOnHostZones] Processing deposit ID:{%d} DENOM:{%s} AMT:{%d}",
+			depositRecord.Id, depositRecord.Denom, depositRecord.Amount))
+
+		hostZone, hostZoneFound := k.GetHostZone(ctx, depositRecord.HostZoneId)
+		if !hostZoneFound {
+			k.Logger(ctx).Error(fmt.Sprintf("[StakeExistingDepositsOnHostZones] Host zone not found for deposit record {%d}", depositRecord.Id))
+			continue
+		}
 
-			ctx.EventManager().EmitEvent(
-				sdk.NewEvent(
-					sdk.EventTypeMessage,
-					sdk.NewAttribute("hostZone", hostZone.ChainId),
-					sdk.NewAttribute("newAmountStaked", strconv.FormatInt(depositRecord.Amount, 10)),
-				),
-			)
+		delegateAccount := hostZone.GetDelegationAccount()
+		if delegateAccount == nil || delegateAccount.GetAddress() == "" {
+			k.Logger(ctx).Error(fmt.Sprintf("[StakeExistingDepositsOnHostZones] Zone %s is missing a delegation address!", hostZone.ChainId))
+			continue
 		}
+
+		k.Logger(ctx).Info(fmt.Sprintf("\t[StakeExistingDepositsOnHostZones] Staking %d on %s", depositRecord.Amount, hostZone.HostDenom))
+		stakeAmount := sdk.NewCoin(hostZone.HostDenom, sdk.NewInt(depositRecord.Amount))
+
+		err := k.DelegateOnHost(ctx, hostZone, stakeAmount, depositRecord.Id)
+		if err != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("Did not stake %s on %s | err: %s", stakeAmount.String(), hostZone.ChainId, err.Error()))
+			continue
+		} else {
+			k.Logger(ctx).Info(fmt.Sprintf("Successfully submitted stake for %s on %s", stakeAmount.String(), hostZone.ChainId))
+		}
+
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				sdk.EventTypeMessage,
+				sdk.NewAttribute("hostZone", hostZone.ChainId),
+				sdk.NewAttribute("newAmountStaked", strconv.FormatInt(depositRecord.Amount, 10)),
+			),
+		)
 	}
 }
 
 func (k Keeper) TransferExistingDepositsToHostZones(ctx sdk.Context, epochNumber uint64, depositRecords []recordstypes.DepositRecord) {
 	transferDepositRecords := utils.FilterDepositRecords(depositRecords, func(record recordstypes.DepositRecord) (condition bool) {
```

Comment:
Great call out, I hadn't thought about that. 

I think `depositRecords` is a copy because the state doesn't actually store a list data type. `GetAllDepositRecords` is iterating through every item in the store, unmarshalling the deposit record into a new object, and building a list from those objects. So by the time the list is returned, the actual state should be no longer referenced, and I think the list should just be a copy.  Which would mean, when we remove elements from the underlying store, it doesn't impact the list. 

I think the case that would be problematic would be if in hooks we were using our own iterator and iterating at the start of each for loop.